lib/semantics/tools.h

```diff
@@ -97,9 +96,39 @@ const Symbol *FindExternallyVisibleObject(
       expr.u);
 }
 
+using SomeExpr = evaluate::Expr<evaluate::SomeType>;
+
 bool ExprHasTypeCategory(
-    const evaluate::GenericExprWrapper &expr, const common::TypeCategory &type);
+    const SomeExpr &expr, const common::TypeCategory &type);
 bool ExprTypeKindIsDefault(
-    const evaluate::GenericExprWrapper &expr, const SemanticsContext &context);
+    const SomeExpr &expr, const SemanticsContext &context);
+
+struct GetExprHelper {
+  const SomeExpr *Get(const parser::Expr::TypedExpr &x) {
+    return x ? &x->v : nullptr;
+  }
+  const SomeExpr *Get(const parser::Expr &x) { return Get(x.typedExpr); }
+  const SomeExpr *Get(const parser::Variable &x) { return Get(x.typedExpr); }
+  template<typename T> const SomeExpr *Get(const common::Indirection<T> &x) {
+    return Get(x.value());
+  }
+  template<typename T> const SomeExpr *Get(const std::optional<T> &x) {
+    return x.has_value() ? Get(x.value()) : nullptr;
+  }
+  template<typename T> const SomeExpr *Get(const T &x) {
+    if constexpr (ConstraintTrait<T>) {
+      return Get(x.thing);
+    } else if constexpr (WrapperTrait<T>) {
+      return Get(x.v);
+    } else {
+      return nullptr;
+    }
+  }
+};
+
+template<typename T> const SomeExpr *GetExpr(const T &x) {
```

Comment:
Idea: this could be a member template function of `SemanticsContext` and it could ensure on failure that the user program is already known to have fatal errors.

Reply:
What about introducing a special value of evaluate::Expr (or of GenericExprWrapper) to indicate the expression is erroneous? Then GetExpr could assert that every typedExpr has a value.

Reply:
We wouldn't want to put an error recovery token into `Expr<>`, it would add an alternative that would have to be checked for and propagated throughout the expression representation and folding code. Maybe `GenericExprWrapper` or as a flag next to the `typedExpr` member(s) in the parse tree.

Reply:
I'll do this in a separate PR.